Windows IORead: return the actual report size

jloehr opened the pull request:
This is kind of Voodoo PR as it intends to fix an issue with the DolphinBar on Windows and i don't have a DolphinBar to reproduce the issue.

The issue that this PR might fix is that sometimes Wiimotes don't work ingame when connected via DolphinBar (i.e. the game does not receive any input). When this occurs the battery level in the Wii Menu will show it's empty regardless of the actual battery level. (See https://forums.dolphin-emu.org/Thread-dolphin-bar-and-windows-10 & https://forums.dolphin-emu.org/Thread-dolphinbar-wiimote-does-not-work).

As @leoetlino stated in one of the threads he encountered the issue on Windows, whereas he didn't on Linux. One difference between Linux and Windows Wiimote LL handling is that on Windows the IORead function reports the wrong amount of bytes read (it is always 23 bytes => MAX_PAYLOAD). If a smaller report is received the remaining/unused bytes are filled with garbage or bytes from previous reports. It may be the case that sometimes those bytes contain certain values that do get interpreted.

_As this is a Voodoo PR let's wait until someone can actually confirm if this fixes the issue_

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4128)

<!-- Reviewable:end -->

mbc07 commented:
I have a Dolphin Bar but I've never faced the issues mentioned on those threads. Is there some specific scenario you would like me to test with this PR?

GitToTheHub commented:
You know, that the links you posted, handle problems with third party wii motes? Third Party Wii-Motes did currently not work with dolphin. Original Nintendo Wii-Motes are working fine with the dolphin bar.

jloehr commented:
leolam said that his third-party Wiimote works on Linux with the DolphinBar, but doesn't on Windows. Additionally one user tested his third-party Wiimote connected via DolphinBar with my test program and it was able to write/read to/from the Wiimote.
So it is apparently a software problem, though it may be a issue within the Windows driver stack.

Regarding this PR it seems to not fix the issue.

GitToTheHub commented:
Ah ok, sorry, i didn't know this pr should be resolving that. It would be great if third party wii motes would function with dolphin, so a lot of money could be saved :) 

I saw you participant on [pr-4090 WiimoteReal: Add a hidapi IO implementation](https://github.com/dolphin-emu/dolphin/pull/4090). Is that pr related to this pr? It's too bad that this pr cannot fix that :(

mbc07 commented:
I'm afraid I'm not of much help there; although I own a Dolphin Bar, all of my Wiimotes are genuine, from Nintendo...

leoetlino commented:
Hmm, too bad this doesn't fix the random issues. Doesn't help that it's tricky to reproduce, as it sometimes works fine.

@GitToTheHub, no, #4090 doesn't change anything significant on Windows. Its main purpose is adding DolphinBar to Linux, while removing/moving duplicate code into common code.

jloehr commented:
> Ah ok, sorry, i didn't know this pr should be resolving that. It would be great if third party wii motes would function with dolphin, so a lot of money could be saved :)

Some do work with Dolphin, they are just not officially supported as some do not.

> I saw you participant on pr-4090 WiimoteReal: Add a hidapi IO implementation. Is that pr related to this pr? It's too bad that this pr cannot fix that :(

Nope, not at all. That PR is only affecting Linux, macOS and i think Android. Though it adds support for DolphinBar on those systems.

leoetlino commented:
@mbc07 so far it seems that third party Wiimotes are most likely to have issues, though there's a forum user who states that this happens for them when more than two (official) Wiimotes are actively used.

jloehr commented:
@leoetlino i think there are two different issues, though they may be related. One is that no input is received from the Wiimote, resulting in an empty battery display. The other is the random disconnect.

jloehr commented:
The latter one is occurring to me with two official Wiimotes (TR and non-TR) connected via BT Dongle.

mbc07 commented:
Hmm, I have two genuine Wiimotes (both non-TR, one is the older "launch" Wiimote with the Motion Plus add-on and the other is an earlier Wii Remote Plus, the non-TR one). I'll do some tests with them and report back. As far as I remember I hit this random disconnect issue one or twice with Donkey Kong Country Returns in the past...

GitToTheHub commented:
@leoetlino Thanks for explanation. Than i understand the purpose of the pr.
@jloehr I didn't know, that currenty some non-nintendo wii motes do work. In the past i tested a third party wii mote with the dolphin bar. The wii mote did successfully connect to the dolphin bar, but as in the forum mentioned, i could not use it in dolphin.

Some issues had too with original wii motes when 4 wii motes were connected:
- Mario Strikers: Sometimes a character of a player is automatically running to left or right though the player is nothing pressed. The player must reconnect the nun chuk to resolve that issue.
- Monopoly: Here i had random disconnects to sometimes

mbc07 commented:
Alright, I had no issues after using this PR for a while. So, at least on Windows 10 with MS BT Stack or Dolphin Bar, while using genuine Wiimotes, there's no issues, no random disconnects, nothing, everything working (except for Wiimote Speaker but it's the same on master since my remotes are non-TR ones)...

jloehr commented:
If no one objects, i would like to close this PR. It doesn't fix the issue, nor does it seem have any other beneficial effect.

GitToTheHub commented:
It does not fix anything but is it not more accurate code? The next one will not think this could be an issue if he see the old code.

shuffle2 commented:
@jloehr It would be nice to use symbolic constants for the switch cases. Since this PR doesn't seem to impact operational correctness and is aiming to clarify what's going on...seems bad to have a switch with magic constants in it.

GitToTheHub commented:
That's true, i don't like the magic constants too. Named constants variables is a good idea or it could be documented why for example a 6 is returned for a 0x20 report-id. From where do you have the knowledge?

jloehr commented:
Changed it to use the stuff from the Wiimote Emulation (though seems those struct are not used anymore). Also added minor comment with link to the Wiibrew Wiki, where the reports and their sizes are listed.

lioncash commented on Source/Core/Core/HW/WiimoteReal/IOWin.cpp:
```diff
@@ -576,15 +577,45 @@ bool WiimoteWindows::IsConnected() const
 }
 
+// See http://wiibrew.org/wiki/Wiimote for the Report IDs and its sizes
+DWORD GetReportSize(u8 report_id)
```
This should likely have `size_t` as its return type, since that's what `sizeof` returns. Calling code can then cast it if necessary.

jloehr replied:
Done

leoetlino commented on Source/Core/Core/HW/WiimoteReal/IOWin.cpp:
```diff
@@ -623,7 +654,18 @@ int IORead(HANDLE& dev_handle, OVERLAPPED& hid_overlap_read, u8* buf, int index)
 
-  return bytes + 1;
+  // ReadFile will always return 22 bytes read.
+  // So we need to calculate the acutal report size by its report id
```
There are some minor typos here; it should be "_actual_ report size by its report _ID_"

jloehr replied:
Done

leoetlino commented on Source/Core/Core/HW/WiimoteReal/IOWin.cpp:
```diff
@@ -623,7 +654,18 @@ int IORead(HANDLE& dev_handle, OVERLAPPED& hid_overlap_read, u8* buf, int index)
+  DWORD report_size = static_cast<DWORD>(GetReportSize(buf[1]));
+  if (report_size == 0)
+  {
+    WARN_LOG(WIIMOTE, "Recieved unsupported report %d in Wii Remote %i", buf[1], index + 1);
```
_Received_

jloehr replied:
Done

leoetlino commented on Source/Core/Core/HW/WiimoteReal/IOWin.cpp:
```diff
@@ -623,7 +654,18 @@ int IORead(HANDLE& dev_handle, OVERLAPPED& hid_overlap_read, u8* buf, int index)
+  }
+
+  // 1 Byte for the Data Report Byte, another for the Report ID and the actual report size
+  bytes = std::min(bytes, report_size + 1) + 1;
```
Is there a case where `report_size + 1` can end up being larger than `bytes`? If not, dropping the std::min makes the intent clearer, imo.

jloehr replied:
Good catch. Unless the OS/Driver messes up...but that is highly unlikely. And if so the memory is still valid, it just contains garbage/old data.
